cpp/src/Slice/Parser.h

```diff
@@ -502,6 +475,32 @@ class Container : public virtual SyntaxTreeBase
     std::map<std::string, ContainedPtr> _introducedMap;
 };
 
+// ----------------------------------------------------------------------
+// DataMemberContainer
+// ----------------------------------------------------------------------
+
+class DataMemberContainer : public virtual Container, public virtual Contained
```

Comment:
Any member function can be written as a static helper function, barring visibility. I think having them as member functions instead of helper functions makes them easier to use (you don’t need to pass in extra parameters, or worry about constness), and helps keep the code more organized by grouping them into a class instead of having them dumped as functions in the namespace.

You mean how I didn’t have ‘Operation’ subclass ‘DataMemberContainer’?
That’s because Operation isn’t a data member container. It’s a parameter container. The behavior of exceptions, structs, and classes is very similar, so they all subclass DataMemberContainer, but the behavior of Operation is very different from the rest. For instance we keep two lists for operation to keep the return parameters and in parameters separate. None of the other classes need to separate their members like this.

You’re right that the main function is overridden in 2/3 of the subclasses, but in both cases they still call the super-class’s method, class and exception just perform an additional check, then defer to the ‘main’ logic which is in the super class.

I feel like the uses case of subclasses is exactly this, providing a container of logic and data that is common to multiple classes.

Reply:
> You mean how I didn’t have ‘Operation’ subclass ‘DataMemberContainer’?
No, as I wrote:
> you didn't use this DataMemberContainer for operation - you used instead two DataMemberList data members.

I don't see the value of this class that nothing uses as is. Just provide non-member functions instead, which is the preferred pattern in C++. Member functions make sense only when non-member functions can't do it, e.g. because your class encapsulate truly private data members.

Reply:
Even though it isn’t used anywhere currently, looking back over my changes there are places where it would be useful, for instance where we create data members. Currently we do 3 separate checks and casts for each of Class, Exception, and Struct, to call each one’s ‘createDataMember’. With the DataMemberContainer subclass, we wouldn’t need any of this craziness.

Yeah I’m sure Scott Meyers would advocate against using member functions here! But I still feel like the subclass has it’s merits. Even if we used helper methods, we would still have repetitive copies of some of the necessary member functions, like those for exposing ‘_dataMembers’, or the virtual ones they have to implement. Centralizing logic and reducing repetition are also decent guidelines to code for.

It reduces cluttering of the namespace, avoids the copies using a public getter would incur, leads to less complex method signatures, and makes for more organized code. Something minor I’ve noticed is that we have a very bad habit of writing logic in slice2x, even though it already exists in Slice! Since the source files in Slice are so huge, the global helper functions are easily lost. Having them directly in the relevant class avoids this.

I personally think this is all worth using a member function instead of a global nonmember function, even if it is indeed a trade-off.